sw/lib/sw/device/silicon_creator/rom_patch.c

```diff
@@ -0,0 +1,280 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/lib/sw/device/silicon_creator/rom_patch.h"
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include "sw/device/silicon_creator/lib/drivers/otp.h"
+#include "sw/lib/sw/device/base/hardened_memory.h"
+#include "sw/lib/sw/device/runtime/hart.h"
+#include "sw/lib/sw/device/silicon_creator/base/sec_mmio.h"
+#include "sw/lib/sw/device/silicon_creator/rom_print.h"
+#include "sw/lib/sw/device/silicon_creator/sigverify/sigverify.h"
+
+#include "hw/top_darjeeling/sw/autogen/top_darjeeling.h"
+#include "otp_ctrl_regs.h"  // Generated.
+#include "rv_core_ibex_regs.h"
+
+#define IBEX_IBUS_ADDR_MATCHING_REG(i) \
+  (RV_CORE_IBEX_IBUS_ADDR_MATCHING_0_REG_OFFSET + 4 * (i))
+#define IBEX_IBUS_ADDR_EN_REG(i) \
+  (RV_CORE_IBEX_IBUS_ADDR_EN_0_REG_OFFSET + 4 * (i))
+
+#define IBEX_DBUS_ADDR_MATCHING_REG(i) \
+  (RV_CORE_IBEX_DBUS_ADDR_MATCHING_0_REG_OFFSET + 4 * (i))
+#define IBEX_DBUS_ADDR_EN_REG(i) \
+  (RV_CORE_IBEX_DBUS_ADDR_EN_0_REG_OFFSET + 4 * (i))
+
+#define IBEX_IBUS_REMAP_ADDR_REG(i) \
+  (RV_CORE_IBEX_IBUS_REMAP_ADDR_0_REG_OFFSET + 4 * (i))
+#define IBEX_DBUS_REMAP_ADDR_REG(i) \
+  (RV_CORE_IBEX_DBUS_REMAP_ADDR_0_REG_OFFSET + 4 * (i))
+
+#define IBEX_IBUS_REGWEN_REG(i) \
+  (RV_CORE_IBEX_IBUS_REGWEN_0_REG_OFFSET + 4 * (i))
+#define IBEX_DBUS_REGWEN_REG(i) \
+  (RV_CORE_IBEX_DBUS_REGWEN_0_REG_OFFSET + 4 * (i))
+
+#define ROM_PATCH_FIELD(field, shift, length)                              \
+  enum { PATCH_REGION_##field##_SHIFT = (shift) };                         \
+  enum { PATCH_REGION_##field##_LENGTH = (length) };                       \
+  enum {                                                                   \
+    PATCH_REGION_##field##_MASK = (((~0UL) >> (32 - (length))) << (shift)) \
+  };
+
+ROM_PATCH_FIELD(LOCK_VALID, 0, 4)
```

Comment:
This might be replaced by a use of the `https://github.com/lowRISC/opentitan/blob/master/sw/device/lib/base/bitfield.h` library.

Reply:
Good call, I moved the code to the bitfield API and it does look simpler, thanks.